File: internal/sshterm/snippets.go
package sshterm

import (
	"time"

	"HermesDeckX/internal/database"

	"gorm.io/gorm"
)

// SSHSnippet represents a saved command snippet for quick execution.
type SSHSnippet struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	HostID    uint           `gorm:"not null;index" json:"host_id"`
	Name      string         `gorm:"size:100;not null" json:"name"`
	Command   string         `gorm:"type:text;not null" json:"command"`
	SortOrder int            `gorm:"default:0" json:"sort_order"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

// SSHSnippetRepo provides CRUD operations for command snippets.
type SSHSnippetRepo struct {
	db *gorm.DB
}

// NewSSHSnippetRepo creates a new snippet repository.
func NewSSHSnippetRepo() *SSHSnippetRepo {
	return &SSHSnippetRepo{db: database.DB}
}

// List returns all snippets for a given host, ordered by sort_order.
func (r *SSHSnippetRepo) List(hostID uint) ([]SSHSnippet, error) {
	var list []SSHSnippet
	if err := r.db.Where("host_id = ?", hostID).Order("sort_order asc, id asc").Find(&list).Error; err != nil {
		return nil, err
	}
	return list, nil
}

// Create inserts a new snippet.
func (r *SSHSnippetRepo) Create(s *SSHSnippet) error {
	return r.db.Create(s).Error
}

// Update modifies an existing snippet.
func (r *SSHSnippetRepo) Update(s *SSHSnippet) error {
	return r.db.Save(s).Error
}

// Delete soft-deletes a snippet by ID.
func (r *SSHSnippetRepo) Delete(id uint) error {
	return r.db.Delete(&SSHSnippet{}, id).Error
}

File: web/services/snippets.ts
// SSH command snippets REST API service layer
import { get, post, put, del } from './request';

export interface SSHSnippet {
  id: number;
  host_id: number;
  name: string;
  command: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface SSHSnippetCreateReq {
  host_id: number;
  name: string;
  command: string;
  sort_order?: number;
}

export interface SSHSnippetUpdateReq {
  name: string;
  command: string;
  sort_order?: number;
}

export const snippetsApi = {
  list: (hostId: number) =>
    get<SSHSnippet[]>(`/api/v1/ssh/snippets?hostId=${hostId}`),
  create: (data: SSHSnippetCreateReq) =>
    post<SSHSnippet>('/api/v1/ssh/snippets', data),
  update: (id: number, data: SSHSnippetUpdateReq) =>
    put<SSHSnippet>(`/api/v1/ssh/snippets?id=${id}`, data),
  delete: (id: number) =>
    del<{ deleted: boolean }>(`/api/v1/ssh/snippets?id=${id}`),
};
